django/contrib/auth/forms.py

```diff
@@ -212,14 +219,32 @@ def clean(self):
         password = self.cleaned_data.get('password')
 
         if username is not None and password:
+            self.pre_authenticate(username)
             self.user_cache = authenticate(self.request, username=username, password=password)
             if self.user_cache is None:
+                # Cache failed login with delay.
+                cache.set(
+                    '%s:%s' % (self.FAILED_LOGIN_CACHE_PREFIX, sha256(force_bytes(username)).hexdigest()),
+                    timezone.now() + timedelta(seconds=self.DELAY_AFTER_FAILED_LOGIN),
+                    60
+                )
```

Comment:
The default cache backend is `LocMemCache`, so this will be per-process, right? Still better than no rate limiting at all, I guess. What is the reason for not using the database? Performance?